File: second semester/homework2_task4/Homework2_task4/Homework2_task4/Calculator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace NameSpaceCalculator
{
    using GeneralStack;
    class Calculator
    {
        /// <summary>
        /// Constructor for calculator.
        /// </summary>
        /// <param name="stack"> Stack, which we will use.</param>
        public Calculator(GeneralStack stack)
        {
            this.stack = stack;
        }

        private GeneralStack stack;

        /// <summary>
        /// Get value.
        /// </summary>
        /// <returns></returns>
        public int PopValue()
        {
            return stack.Pop();
        }

        /// <summary>
        /// Add result.
        /// </summary>
        /// <param name="value">Value to be pushed.</param>
        public void PushResult(int value)
        {
            stack.Push(value);
        }

        /// <summary>
        /// Addition.
        /// </summary>
        public void Addition()
        {
            PushResult(stack.Pop() + stack.Pop());
        }

        /// <summary>
        /// Subtraction.
        /// </summary>
        public void Subtraction()
        {
            PushResult(stack.Pop() - stack.Pop());
        }

        /// <summary>
        /// Multiplication.
        /// </summary>
        public void Multiplication()
        {
            PushResult(stack.Pop() * stack.Pop());
        }

        /// <summary>
        /// Division.
        /// </summary>
        public void Division()
        {
            PushResult(stack.Pop() / stack.Pop());
        }
    }
}

File: QT/calculator/calculator.cpp
#include "calculator.h"

Calculator::~Calculator()
{}

void Calculator::operation(QCharRef something)
{
    if (something == '+')
    {
        this->pushResult(this->popValue() + this->popValue());
    }

    if (something == '-')
    {
        this->pushResult(this->popValue() - this->popValue());
    }

    if (something == '*')
    {
        this->pushResult(this->popValue() * this->popValue());
    }

    if (something == '/')
    {
        int const number1 = this->popValue();
        int const number2 = this->popValue();
        if (number1 == 0)
            throw QString("Division by 0");
        this->pushResult(number2 / number1);
    }
}

int Calculator::popValue()
{
    return this->stack.pop();
}

void Calculator::pushResult(int value)
{
    this->stack.push(value);
}

void Calculator::removeCalculator(GeneralStack &stack)
{
    this->stack.removeStack();
}
